### ISSUE TITLE
NgUpgrade for Performance's `downgradeModule` can only be called once

### ISSUE THREAD

**Issue by rkirov** [angular/angular#26062]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[ ] Feature request
[x] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
A call like:

  `module('demoApp', [downgradeModule(Ng2Mod1), downgradeModule(Ng2Mod2)])`

succeeds but only the first module is usable. 

## Expected behavior

Either both modules are available or an error is throw describing the limitation. The workaround is not too bad IMHO, just import Ng2Mod1 and Ng2Mod2 into a single module, but without an error this is confusing to users.

## Minimal reproduction of the problem with instructions
https://stackblitz.com/edit/ngupgradelite-multi-downgrademodule?file=ng1%2Fdemo-app.module.ts

</code></pre>

**Comment by gkalpak** [comment]
Actually, if you [set everything up correctly][1] (downgrade all components, etc), you do get an error:

> Unhandled Promise rejection:
> No component factory found for TestComponent. Did you add it to @NgModule.entryComponents? ; Zone: <root> ; Task: Promise.then ; Value:
>   Error: No component factory found for TestComponent. Did you add it to @NgModule.entryComponents?

This is because in AngularJS registering a module silently overwrites any previously registered module with the same name. Thus, the last `downgradeModule()` call takes effect and the component tesolver fails when trying to instantiate any component from the first downgraded module.

##
This is just for clarification of what is going on. I do agree that a more explicit and actionable error message is needed.

[1]: https://stackblitz.com/edit/ngupgradelite-multi-downgrademodule-ymzfmr?file=ng1/demo-app.module.ts

**Comment by jrabbe** [comment]
We are running into this error as we are investigating and prototyping the migration from AngularJS to Angular.

Our we application is running AngularJS 1.6.6, and we have been moving towards a modular approach as suggested by the [Angular Styleguide](https://github.com/johnpapa/angular-styleguide/tree/master/a1#many-small-self-contained-modules). Before using modules our app relied heavily on lazy-loading, so we have started using [ocLazyLoad](https://oclazyload.readme.io/) to lazy load modules as needed.

The Angular prototype I have created uses ocLazyLoad to load an Angular module for a route downgraded using `downgradeModule`. This works and all is well. However, when I introduce a downgraded Angular module for a second route, is where the trouble starts. When I go to the first route the downgraded module loads, but when I got to the second route it won't load the second module. 

Do to the way I am loading the downgraded modules I create an AngularJS module using the downgraded module as a dependency, and registering the downgraded component as a directive.

```
import angular from 'angular';
import {downgradeComponent, downgradeModule} from '@angular/upgrade/static';
import {Page1Module} from './page1.module';
import {Page1Component} from './page1.component';

angular.module('page1.module', [downgradeModule(Page1Module)])
  .directive('page1', {component: Page1Component}
```

I have a similar definition for my page 2, and through my UI router setup I load the module corresponding to the state using a `resolve` function, and set the `component` as the page component.

I will try and set up a blitz to demonstrate the setup, though it's more complicated than the example provided above.

**Comment by jrabbe** [comment]
Blitz showing our setup: https://stackblitz.com/edit/ngupgradelite-multi-downgrademodule-fbqapd

To reproduce:

1. Click on either of the "Test 1" or "Test 2" links
1. Click the "Back" link
1. Click the link not clicked in step 1 above.

**Comment by gkalpak** [comment]
I don't think there is a way to automatically (and reliably) map downgraded component to downgraded modules, but it should be possible to make this setup work by manually specifying which module each downgraded component belongs to. E.g. something like:

```ts
const downMod1 = downgradeModule(bootstrapFn1);
const downMod2 = downgradeModule(bootstrapFn2);

angular.
  module('ng1', [downMod1, downMod2]).
  directive('compA', downgradeComponent({component: ComponentA, module: downMod1}).
  directive('compB', downgradeComponent({component: ComponentB, module: downMod2});
```

Would something like this work for you, @jrabbe?

**Comment by jrabbe** [comment]
I don't have a problem manually mapping components to modules for downgrading purposes.

**Comment by jenniferfell** [comment]
@gkalpak Feel free to triage this differently.

**Comment by gkalpak** [comment]
Here are the two alternatives:
- Better throwing: #26062
- Support for multiple downgraded modules: #26217

**Comment by jrabbe** [comment]
Seems like #26084 is the PR for better throwing, #26062 is the number of this issue.

#26217 looks like it would solve my use case, and would solve the use cases where you are migrating multiple different parts of an AngularJS application. 

I am unsure what the value of #26084 is. If you can't downgrade multiple modules you would need to have one root Angular module to downgrade, and depend on it at the earliest common root of the different places of the application you are migrating. Wouldn't you just end up having to depend on the downgraded root Angular module almost from the outset, mitigating most of the benefit in using "Upgrade for performance" and the possibility of delaying the loading of Angular and its dependencies?

**Comment by gkalpak** [comment]
> I am unsure what the value of #26084 is.

If we decide that we don't want (or can't) support multiple downgraded modules, then #26084 improves the reported error.

> If you can't downgrade multiple modules [...] wouldn't you just end up having to depend on the downgraded root Angular module almost from the outset [...] mitigating most of the benefit in using "Upgrade for performance"?

It is possible to do lazy-loading in AngularJS, so you could still delay loading Angular and its dependencies until the user visit a part of the app where it is needed. But you would basically need to load all of it once you hit the first point in the app where you need something from Angular-land.
(For the record, "Upgrade for performance" has other significant benefits, which you an still enjoy (e.g. the decoupled change detections).

##
Basically, I created #26084 as a quick improvement over the current state (inconsistent and cryptic error), but then I realized it wouldn't be much more difficult to properly support multiple downgraded modules (which is something that wanted to do anyway at some point :grin), so I started working on #26217 (and then it took me two weeks to finish it 😅 🤷‍♂️).

Anyway, I think #26217 is the way to go 💪

**Comment by jrabbe** [comment]
> If we decide that we don't want (or can't) support multiple downgraded modules, then #26084 improves the reported error.

Oh… that makes sense.

> It is possible to do lazy-loading in AngularJS, so you could still delay loading Angular and its dependencies until the user visit a part of the app where it is needed.

We are currently using lazy-loading in AngularJS, the main drawback is that we might need to load much more than we need once we need to load Angular… I don't think I made that point very clearly.

> Anyway, I think #26217 is the way to go 💪

Awesome work, thank you.

**Comment by jrabbe** [comment]
Just posted a question on #26217: [this is unfortunately blocking us](https://github.com/angular/angular/pull/26217#issuecomment-434391358) right now.
